8663 support job cancellation

ttoomey opened the pull request:
## _Merging this PR_
- use the squash-merge strategy for PRs targeting a release-X branch

### Description
Adds a job cancellation mechanism allowing administrators to cancel running delayed jobs 

### Functional Areas
- Added `cancel` action and UI controls to the delayed jobs dashboard.
- Added `handle_cancellation!` in `ApplicationJob` based on `cancellation_requested_at` timestamp.
- Added cancellation checks during report checkpoint progress tracking to abort execution if requested.
- Fixed up delayed job index page view
- Instrumented the HUD report runner and SPM enrollment filters with cancellation checkpoints to support mid-run stopping.
- interruptible jobs check for stopping workers and raise to exit and reschedule themselves
- Adds a dj plugins to track workers, enabling detection of stopping workers
- Corrects infinite loop in DJ processing in CI

### Type of Change
New feature

## Checklist before requesting review
[//]: # (Remove any items that are not applicable)
- [x] I have performed a self-review of my code
- [x] I have run the code that is being changed under ideal conditions, and it doesn't fail
- [x] I have updated the documentation (or not applicable)
- [x] I have added spec tests (or not applicable)
- [x] I have provided testing instructions in this PR or the related issue (or not applicable)

ttoomey commented on app/views/admin/delayed_jobs/_row.haml:
```diff
@@ -0,0 +1,63 @@
+      = job.cancellation_requested_at
+  %td= job.queue
+  %td= job.priority
+  - if can_manage_config?
```
this is new, hiding the error for non-admins

eanders commented on app/jobs/application_job.rb:
```diff
@@ -7,4 +7,19 @@
+    # We can't access the job record directly from the instance easily without
+    # looking it up.
+    job = Delayed::Job.find_by(id: provider_job_id)
+    job&.handle_cancellation!
```
I'm probably missing something, but isn't this a circular call?
I see it's also defined in the monkey patch, so maybe that's what get's called.  If so, it looks like it raises, which maybe would be caught by `discard_on JobCancelled`.  We don't have the discard gem, but is `discard_on` part of delayed job?  How is this actually stopping the worker?

Maybe some descriptive documentation would be helpful.

ttoomey replied:
So this is in ActiveJob::Base; it delegates `handle_cancellation!` to the delayed job model which is where the actual cancellation logic and state live. Because these are different models, the call is not circular. 

The `discard_on` directive in active job is not related to the `discard` gem but actually part of the standard rails active job API. [See docs](https://guides.rubyonrails.org/active_job_basics.html#retrying-or-discarding-failed-jobs)

I'll add some documentation to help explain

eanders replied:
Thanks!  
I think some docs will make this more clear, and nice to know about `discard_on`.

eanders commented on app/models/hud_reports/report_instance.rb:
```diff
@@ -168,6 +168,8 @@ def start_report
     def track_progress(checkpoint_name)
       raise unless checkpoint_name.present?
 
+      related_job&.handle_cancellation!
```
Ok, so this is what will end up raising, which makes sense, that could kill the worker, but it looks like it's only implemented for HUD reports, but the option to cancel might show on all jobs.

ttoomey replied:
It shouldn't kill the worker, if I understand the code, it seems to rescue from the exception ([ref](https://github.com/rails/rails/blob/efe7f5241b751a70cccc86306fb659ee7c4094e2/activejob/lib/active_job/exceptions.rb#L109)).

However we can add the distinction between cancellable vs interruptible. With this change, all jobs can be cancelled before they start (due to `before_perform`) but only hud report jobs are interruptible. We could extend this to other jobs types to support cancellation

eanders replied:
That's a good distinction (the cancel the job rather than kill the worker.)  
I think this is fine as is.  In the future we should extend it to:
1. All HUD reports
2. Importers (can't tell you how many times people have asked us to stop an upload they just uploaded.)
3. Any reports that take a long time to run

ttoomey committed:
squash 8663-support-job-cancellation

ttoomey committed:
pr feedback and cleanup

ttoomey committed:
further enhancements, support interruptable vs cancellable

ttoomey committed:
enhancement

ttoomey committed:
cleanup

ttoomey committed:
cleanups

ttoomey committed:
Merge branch 'release-192' into 8663-support-job-cancellation

ttoomey committed:
fixes

ttoomey commented:
changes since last review:

- **Job Management Dashboard**: Adds concurrency-safe actions and guards in the controller
- **HUD SPM Report**:  check for interruption during enrollment generation.
- **Job Metadata Extraction**: docs and support for varying payload formats (ActiveJob vs. native Delayed Job).
- **Admin Performance**: Optimization for batch loading only required users 
- ** Job Execution **: supports interruptible? execution

ttoomey committed:
test fix

ttoomey committed:
handle sigterm

ttoomey commented:
@eanders 
considering additional enhancement to this branch; I feel like it would be good to also trap sigterm and then retry the job; this would allow long-running jobs to stop gracefully when the pod is evicted. I believe this would help the case where a job needs more memory, VPA tries to give it more, but the job keeps running in a degraded state rather than stopping on sigterm

ttoomey committed:
thread safety

ttoomey committed:
cleanups

ttoomey committed:
tidy, rename method

ttoomey committed:
for testing

ttoomey committed:
cleanups

ttoomey committed:
cleanup

ttoomey committed:
fixes

ttoomey committed:
rubocop

ttoomey committed:
cleanup

ttoomey committed:
Merge pull request #6114 from greenriver/8663-support-job-term

8663 support sig term in jobs

ttoomey committed:
cleanup

ttoomey committed:
Merge branch 'release-192' into 8663-support-job-cancellation

ttoomey committed:
test fix?

ttoomey committed:
extract

eanders commented on spec/models/active_record_extended_merge_regression_spec.rb:
```diff
@@ -14,11 +14,11 @@
 
   before(:all) do
-    setup(default_setup_path)
+    setup(default_setup_path) if ENV['SKIP_AREL_SQL_LITERAL_INITIALIZER'] == '1'
```
Why these changes?

ttoomey replied:
Technically this saves some time time in CI since it runs setup but doesn't run the spec unless the env var is set

But I'll back it out; I was trying to fix a hang in CI that was actually caused by an infinite waiting for the job queue to be empty (fixed with `work_off_all_ready_jobs` helper

ttoomey committed:
pr feedback

ttoomey committed:
test fix?

ttoomey commented:
Looks like still some lingering test failures; my plan is to merge deploy to QA before merging into the main release

ttoomey committed:
Merge branch 'release-193' into 8663-support-job-cancellation

ttoomey committed:
docs, test fix?

ttoomey committed:
Merge branch 'release-193' into 8663-support-job-cancellation

ttoomey committed:
fix?

ttoomey committed:
refactored SignalHandlerPlugin with Thread.current

ttoomey committed:
nil-checks on dates

ttoomey committed:
Revert "nil-checks on dates"

This reverts commit 3575d094f1606621ee8500f672e2a5ea11bba713.

ttoomey committed:
test fix?

ttoomey committed:
Merge branch 'release-193' into 8663-support-job-cancellation

ttoomey committed:
revert waiter improvement

ttoomey committed:
fixes

ttoomey committed:
cleanup

ttoomey committed:
fix

ttoomey committed:
wip

ttoomey committed:
fix

ttoomey committed:
fix?

ttoomey committed:
Merge branch 'release-194' into 8663-support-job-cancellation

ttoomey committed:
fix?

ttoomey committed:
Merge branch 'release-195' into 8663-support-job-cancellation

ttoomey committed:
Revert "fix?"

This reverts commit d5dc3c84658cdac7c16fa173e65fdc849cac6a87.

ttoomey committed:
fix & tests

ttoomey committed:
8663-support-job-cancellation

ttoomey committed:
test fix again?

ttoomey committed:
Merge branch 'release-195' into 8663-support-job-cancellation

ttoomey committed:
test fix again

ttoomey committed:
fix?

ttoomey committed:
fix

ttoomey committed:
standardize the job re-queue

ttoomey committed:
base job spec

ttoomey commented:
I had to iterate quite a bit to get specs to pass, a few things changed enough to warrant a quick re-review before merge.

eanders commented on spec/jobs/base_job_spec.rb:
```diff
@@ -0,0 +1,103 @@
+
+  describe '#delayed_job' do
+    context 'when provider_job_id is present' do
+      let!(:dj_record) { Delayed::Job.create!(handler: 'dummy') }
+
+      before do
+        allow(job_instance).to receive(:provider_job_id).and_return(dj_record.id)
+      end
+
+      it 'returns the job record using provider_job_id' do
+        expect(job_instance.delayed_job).to eq(dj_record)
+      end
```
What is this testing?  It looks like we mock `provider_job_id` and tell it what to return, then check that the mock worked.  I'm probably missing something.

ttoomey replied:
I think it's just to prove that the `delayed_job` method uses `provider_job_id` to find the record; I don't mind it I guess

eanders commented on spec/support/delayed_job_helpers.rb:
```diff
@@ -0,0 +1,39 @@
+
+      until total_processed >= job_limit
+        # Handling potential clock skew in CI by forcing jobs to be ready
+        Delayed::Job.where(run_at: Time.current..).update_all(run_at: 1.minute.ago)
```
This seems like it could break tests that intentionally queue something in the future and expect it to not be run.  Not sure we have any of those, but if we're messing with time in the test, this could cause problems.

ttoomey replied:
Thanks, I removed it; it's still not totally clear to me what the issue was with this helper; I think the worker may have been trying to consume the wrong queue.

eanders reviewed (commented):
@ttoomey this looks good.  I dropped a few comments I think might need addressing before we merge.

ttoomey committed:
fix

ttoomey committed:
Merge branch 'release-196' into 8663-support-job-cancellation

ttoomey committed:
Revert "fix"

This reverts commit 57a94a0a2e99360ec33f08375e5b8623f06dba95.

ttoomey committed:
Revert "Revert "fix""

This reverts commit 909dfb41fd4695ea3677846939216f5fb9852ed7.

ttoomey committed:
bump timing check for CI
